rcl/test/rcl/test_timer.cpp

```diff
@@ -246,3 +248,223 @@ TEST_F(TestTimerFixture, test_canceled_timer) {
+  });
+
+  ASSERT_EQ(RCL_RET_OK, rcl_set_ros_time_override(&clock, sec_1)) << rcl_get_error_string_safe();
+  ASSERT_EQ(RCL_RET_OK, rcl_enable_ros_time_override(&clock)) << rcl_get_error_string_safe();
+
+  int64_t time_until_pre = 0;
+  ASSERT_EQ(RCL_RET_OK, rcl_timer_get_time_until_next_call(&timer, &time_until_pre)) <<
+    rcl_get_error_string_safe();
+  ASSERT_EQ(sec_5 - (sec_1 - 1), time_until_pre);
+
+  ASSERT_EQ(RCL_RET_OK, rcl_disable_ros_time_override(&clock)) << rcl_get_error_string_safe();
+
+  int64_t time_until = 0;
+  ASSERT_EQ(RCL_RET_OK, rcl_timer_get_time_until_next_call(&timer, &time_until)) <<
+    rcl_get_error_string_safe();
+  // Because of time credit the time until next call should be less than before
+  EXPECT_GT(time_until_pre, time_until);
+  EXPECT_LT(0, time_until);
+}
+
+TEST_F(TestTimerFixture, test_ros_time_backwards_jump) {
+  rcl_ret_t ret;
+  const int64_t sec_5 = 5LL * 1000LL * 1000LL * 1000LL;
+  const int64_t sec_3 = 3LL * 1000LL * 1000LL * 1000LL;
+  const int64_t sec_2 = 2LL * 1000LL * 1000LL * 1000LL;
+  const int64_t sec_1 = 1000LL * 1000LL * 1000LL;
+
+  rcl_clock_t clock;
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  ret = rcl_clock_init(RCL_ROS_TIME, &clock, &allocator);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+    EXPECT_EQ(RCL_RET_OK, rcl_clock_fini(&clock)) << rcl_get_error_string_safe();
+  });
+
+  ASSERT_EQ(RCL_RET_OK, rcl_set_ros_time_override(&clock, sec_2)) << rcl_get_error_string_safe();
+  ASSERT_EQ(RCL_RET_OK, rcl_enable_ros_time_override(&clock)) << rcl_get_error_string_safe();
+
+  rcl_timer_t timer = rcl_get_zero_initialized_timer();
+  ret = rcl_timer_init(&timer, &clock, sec_5, nullptr, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string_safe();
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+    EXPECT_EQ(RCL_RET_OK, rcl_timer_fini(&timer)) << rcl_get_error_string_safe();
+  });
+
+  ASSERT_EQ(RCL_RET_OK, rcl_set_ros_time_override(&clock, sec_3)) << rcl_get_error_string_safe();
+  {
+    // Moved forward a little bit, timer should be closer to being ready
+    int64_t time_until = 0;
+    ASSERT_EQ(RCL_RET_OK, rcl_timer_get_time_until_next_call(&timer, &time_until)) <<
+      rcl_get_error_string_safe();
+    EXPECT_EQ(sec_5 - (sec_3 - sec_2), time_until);
+  }
+  ASSERT_EQ(RCL_RET_OK, rcl_set_ros_time_override(&clock, sec_1)) << rcl_get_error_string_safe();
+  {
+    // Jumped back before timer was created, so last_call_time should be 1 period
+    int64_t time_until = 0;
+    ASSERT_EQ(RCL_RET_OK, rcl_timer_get_time_until_next_call(&timer, &time_until)) <<
+      rcl_get_error_string_safe();
+    EXPECT_EQ(sec_5, time_until);
+  }
+}
+
+TEST_F(TestTimerFixture, test_ros_time_wakes_wait) {
+  const int64_t sec_5 = 5LL * 1000LL * 1000LL * 1000LL;
+  const int64_t sec_1 = 1000LL * 1000LL * 1000LL;
+  const int64_t sec_1_5 = 1500LL * 1000LL * 1000LL;
```

Comment:
For reference there is `RCL_S_TO_NS`:

https://github.com/ros2/rcl/blob/a0f83b071a67b1538f2210a1265fbb814aa0abd7/rcl/include/rcl/time.h#L30

https://github.com/ros2/rcutils/blob/f04cc666dd06347050ff5c18b80079922e729d92/include/rcutils/time.h#L30

Maybe you avoided it to get the "right" literal type to avoid compiler warnings, but just wanted to mention in case you didn't know about it.

Reply:
I forgot about those. The `1000LL` is indeed to avoid compiler warnings (and test failures) on windows. I update this PR to use those when ros2/rcutils#118 is merged.

Reply:
`RCL_S_TO_NS` in 5fc6f10